agent/agent.go

```diff
@@ -1253,6 +1278,133 @@ func (a *agent) startReportingConnectionStats(ctx context.Context) {
+			slog.F("env_var", EnvProcMemNice),
+			slog.F("value", val),
+			slog.F("goos", runtime.GOOS),
+		)
+		return
+	}
+
+	manage := func() {
+		procs, err := a.manageProcessPriority(ctx)
+		if err != nil {
+			a.logger.Error(ctx, "manage process priority",
+				slog.F("dir", agentproc.DefaultProcDir),
+				slog.Error(err),
+			)
+		}
+		if a.modifiedProcs != nil {
+			a.modifiedProcs <- procs
+		}
+	}
+
+	// Do once before falling into loop.
+	manage()
+
+	if a.processManagementTick == nil {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		a.processManagementTick = ticker.C
+	}
+
+	for {
+		select {
+		case <-a.processManagementTick:
+			manage()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (a *agent) manageProcessPriority(ctx context.Context) ([]*agentproc.Process, error) {
+	const (
+		procDir     = agentproc.DefaultProcDir
+		niceness    = 10
+		oomScoreAdj = -500
+	)
+
+	procs, err := agentproc.List(a.filesystem, a.syscaller, agentproc.DefaultProcDir)
+	if err != nil {
+		return nil, xerrors.Errorf("list: %w", err)
+	}
+
+	modProcs := []*agentproc.Process{}
+	for _, proc := range procs {
+		// Trim off the path e.g. "./coder" -> "coder"
+		name := filepath.Base(proc.Name())
+		// If the process is prioritized we should adjust
+		// it's oom_score_adj and avoid lowering its niceness.
+		if slices.Contains(prioritizedProcs, name) {
+			err = proc.SetOOMAdj(oomScoreAdj)
+			if err != nil {
+				a.logger.Error(ctx, "unable to set proc oom_score_adj",
+					slog.F("name", proc.Name()),
+					slog.F("pid", proc.PID),
+					slog.F("oom_score_adj", oomScoreAdj),
+					slog.Error(err),
+				)
+				continue
+			}
+			modProcs = append(modProcs, proc)
+
+			a.logger.Debug(ctx, "decreased process oom_score",
+				slog.F("name", proc.Name()),
+				slog.F("pid", proc.PID),
+				slog.F("oom_score_adj", oomScoreAdj),
+			)
+			continue
+		}
+
+		score, err := proc.Niceness(a.syscaller)
+		if err != nil {
+			a.logger.Error(ctx, "unable to get proc niceness",
+				slog.F("name", proc.Name()),
+				slog.F("pid", proc.PID),
+				slog.Error(err),
+			)
+			continue
+		}
+
+		// We only want processes that don't have a nice value set
+		// so we don't override user nice values.
+		// Getpriority actually returns priority for the nice value
+		// which is niceness + 20, so here 20 = a niceness of 0 (aka unset).
+		if score != 20 {
+			a.logger.Error(ctx, "skipping process due to custom niceness",
+				slog.F("name", proc.Name()),
+				slog.F("pid", proc.PID),
+				slog.F("niceness", score),
+			)
+			continue
+		}
+
+		err = proc.SetNiceness(a.syscaller, niceness)
+		if err != nil {
+			a.logger.Error(ctx, "unable to set proc niceness",
+				slog.F("name", proc.Name()),
+				slog.F("pid", proc.PID),
+				slog.F("niceness", niceness),
+				slog.Error(err),
+			)
+			continue
+		}
+
+		a.logger.Debug(ctx, "deprioritized process",
+			slog.F("name", proc.Name()),
+			slog.F("pid", proc.PID),
```

Comment:
These log fields are duplicated a few times. Should create a sublogger with `a.logger.With`.